docs(snowflake): document transient tmp_relation_type and resolve macro (#9098)
fishtown-analytics/docs.getdbt.com: website/docs/reference/resource-configs/snowflake-configs.md

## Temporary tables

> Incremental table merges for Snowflake prefer to utilize a `view` rather than a `temporary table`. The reasoning is to avoid the database write step that a temporary table would initiate and save compile time.

Reviewer: ```suggestion
To save compile time and avoid the database write step initiated by a temporary table, incremental table merges for Snowflake prefer to utilize a `view` rather than a `temporary table` .
```

> However, some situations remain where a temporary table would achieve results faster or more safely. The `tmp_relation_type` configuration enables you to opt in to temporary or transient tables for incremental builds. This is defined as part of the model configuration.

Reviewer: ```suggestion
Sometimes a temporary table achieves results faster or more safely. You can opt in to temporary or transient tables for incremental builds by using the `tmp_relation_type` configuration  This is defined as part of the model configuration.
```

To guarantee accuracy, an incremental model using the `delete+insert` strategy with a `unique_key` defined requires a temporary table; trying to change this to a view will result in an error.

> `tmp_relation_type` accepts three values:

Reviewer: Let's not number in case we have more later.
```suggestion
`tmp_relation_type` accepts these values:
```

> - `view` (default): No database write for the tmp relation; fastest but not suitable for all strategies.

Reviewer: Just want to clarify "no database write"
```suggestion
- `view` (default): Skips intermediate step of creating a temporary physical table for the tmp relation; fastest but not suitable for all strategies.
```

- `table`: A session-scoped temporary table; not visible in the Snowflake catalog and isolated per session.

> - `transient`: A transient table; persists in the catalog, enabling [Snowflake native lineage tracking](https://docs.snowflake.com/en/user-guide/ui-snowsight-lineage), while avoiding the 7-day fail-safe storage costs of permanent tables. Supported with `delete+insert` and `microbatch` strategies.

Reviewer: OK weird that GH didn't take this change but seemed to
```suggestion
- `transient`: A transient table; persists in the catalog, enabling [Snowflake native lineage tracking](https://docs.snowflake.com/en/user-guide/ui-snowsight-lineage), while avoiding the 7-day fail-safe storage costs of permanent tables.
  **Note:** This value is distinct from the separate model-level `transient` config described later, which controls the final model relation. 
```

> - `transient`: A transient table; persists in the catalog, enabling [Snowflake native lineage tracking](https://docs.snowflake.com/en/user-guide/ui-snowsight-lineage), while avoiding the 7-day fail-safe storage costs of permanent tables. Supported with `delete+insert` and `microbatch` strategies.

Reviewer: Do we want to say this is ideal to use with delete+insert and microbatch strategies? Why?
```suggestion
- `transient`: A transient table; persists in the catalog, enabling [Snowflake native lineage tracking](https://docs.snowflake.com/en/user-guide/ui-snowsight-lineage), while avoiding the 7-day fail-safe storage costs of permanent tables.
  **Note:** This value is distinct from the separate model-level `transient` config described later, which controls the final model relation. 
```

Author: I'm actually not too sure about why we have the comment "Supported with `delete+insert` and `microbatch` strategies."

Let me check quickly if we need this part or not.

Author: OK, so trasient works for all, we could then get

- `transient`: A transient table; persists in the catalog, enabling [Snowflake native lineage tracking](https://docs.snowflake.com/en/user-guide/ui-snowsight-lineage), while avoiding the 7-day fail-safe storage costs of permanent tables.
  **Note:** This value is distinct from the separate model-level `transient` config described later, which controls the final model relation.

Reviewer: removed the the delete+insert comment

<File name='dbt_model.sql'>

```sql

{{ config(

>     tmp_relation_type="table | view | transient", ## If not defined, view is the default.

Reviewer: ```suggestion
    tmp_relation_type="table | view | transient", 
    -- If not defined, view is the default.
```

) }}

```

</File>

:::warning Concurrent run conflicts with `transient`

> When `tmp_relation_type` is set to `transient`, the tmp relation is a real table that persists in the target schema under a deterministic name. If multiple runs of the same incremental model execute concurrently in the same schema — for example, when developers share a target schema or when CI and production runs overlap — they can overwrite each other's tmp relation, causing data duplication or incorrect results.

Reviewer: ```suggestion
When `tmp_relation_type` is set to `transient`, the tmp relation becomes a real table that persists in the target schema under a deterministic name. If multiple runs of the same incremental model execute concurrently in the same schema, they can overwrite each other's tmp relation, causing data duplication or incorrect results. For example, this might happen when developers share a target schema or when CI and production runs overlap.
```

> This risk depends on how schemas and databases are configured across developers and environments. To prevent conflicts, use [`snowflake__resolve_incremental_tmp_relation`](#avoiding-tmp-relation-conflicts) to route tmp relations to a schema that is unique per run or environment.

Reviewer: ```suggestion
This risk depends on how you configure schemas and databases  for your dbt models. To prevent conflicts, use `snowflake__resolve_incremental_tmp_relation` to route tmp relations to a schema that is unique per run or environment. For more information, refer to [Avoiding tmp relation conflicts](#avoiding-tmp-relation-conflicts).
```

Author: This risk depends on how you configure schemas and databases for your dbt models.

Reviewer: added!

:::

### Avoiding tmp relation conflicts

To prevent name collisions across concurrent runs, override the `snowflake__resolve_incremental_tmp_relation` dispatch macro to redirect the tmp relation to a dedicated schema:

<File name='macros/snowflake_incremental.sql'>

```sql
{% macro snowflake__resolve_incremental_tmp_relation(tmp_relation) %}
  {{ return(tmp_relation.incorporate(schema='scratch')) }}
{% endmacro %}
```

</File>

> This macro receives the default tmp relation object and returns a modified version. Common overrides include appending a developer username, a CI job ID, or a target name to the schema to ensure isolation across concurrent runs. For example:

Reviewer: ```suggestion
This macro receives the default tmp relation object and returns a modified version. Common overrides include appending a developer username, a CI job ID, or a target name to the schema to ensure isolation across concurrent runs. 

To append a target name to the schema:
```

<File name='dbt_model.sql'>

```sql

{{ config(
    tmp_relation_type="table | view | transient", 
    -- If not defined, view is the default.
) }}

```

</File>

:::warning Concurrent run conflicts with `transient`

When `tmp_relation_type` is set to `transient`, the tmp relation becomes a real table that persists in the target schema under a deterministic name. If multiple runs of the same incremental model execute concurrently in the same schema, they can overwrite each other's tmp relation, causing data duplication or incorrect results. For example, this might happen when developers share a target schema or when CI and production runs overlap.

> This risk depends on how you configure schemas and databases  for your dbt models. To prevent conflicts, use `snowflake__resolve_incremental_tmp_relation` to route tmp relations to a schema that is unique per run or environment. For more information, refer to [Avoiding tmp relation conflicts](#avoiding-tmp-relation-conflicts).

Reviewer: ```suggestion
This risk depends on how you configure schemas and databases for your dbt models. To prevent conflicts, use `snowflake__resolve_incremental_tmp_relation` to route tmp relations to a schema that is unique per run or environment. For more information, refer to [Avoiding tmp relation conflicts](#avoiding-tmp-relation-conflicts).
```
